Support Kwargs in C++ Function/Method calls

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #19217 Make it possible for self.forward to return a ScriptMethod
* #19142 graph_for based on last_optimized_executed_graph
* #19141 Remove GraphExecutor's python bindings
* #19138 Cleanup ScriptModule bindings
* #19137 get propagate_shape logic out of module.h
* #19136 Make debug subgraph inlining thread local
* **#19086 Support Kwargs in C++ Function/Method calls**

Summary: Adds a Kwargs optional argument to the operator() for Method
and Function. Refactors logic from the python support for kwargs so
that it can be shared

Test Plan: test_jit.py TestJit.test_cpp

Differential Revision: [D14875729](https://our.internmc.facebook.com/intern/diff/D14875729)

zdevito committed:
Support Kwargs in C++ Function/Method calls

Summary: Adds a Kwargs optional argument to the operator() for Method
and Function. Refactors logic from the python support for kwargs so
that it can be shared

Test Plan: test_jit.py TestJit.test_cpp

zdevito committed:
Update on "Support Kwargs in C++ Function/Method calls"

Support Kwargs in C++ Function/Method calls

Summary: Adds a Kwargs optional argument to the operator() for Method
and Function. Refactors logic from the python support for kwargs so
that it can be shared

Test Plan: test_jit.py TestJit.test_cpp

gh-metadata: pytorch pytorch 19086 gh/zdevito/13/head

zdevito committed:
Update on "Support Kwargs in C++ Function/Method calls"

Support Kwargs in C++ Function/Method calls

Summary: Adds a Kwargs optional argument to the operator() for Method
and Function. Refactors logic from the python support for kwargs so
that it can be shared

Test Plan: test_jit.py TestJit.test_cpp

gh-metadata: pytorch pytorch 19086 gh/zdevito/13/head

zdevito committed:
Update on "Support Kwargs in C++ Function/Method calls"

Support Kwargs in C++ Function/Method calls

Summary: Adds a Kwargs optional argument to the operator() for Method
and Function. Refactors logic from the python support for kwargs so
that it can be shared

Test Plan: test_jit.py TestJit.test_cpp

gh-metadata: pytorch pytorch 19086 gh/zdevito/13/head

zdevito committed:
Update on "Support Kwargs in C++ Function/Method calls"

Support Kwargs in C++ Function/Method calls

Summary: Adds a Kwargs optional argument to the operator() for Method
and Function. Refactors logic from the python support for kwargs so
that it can be shared

Test Plan: test_jit.py TestJit.test_cpp

gh-metadata: pytorch pytorch 19086 gh/zdevito/13/head

zdevito committed:
Update on "Support Kwargs in C++ Function/Method calls"

Support Kwargs in C++ Function/Method calls

Summary: Adds a Kwargs optional argument to the operator() for Method
and Function. Refactors logic from the python support for kwargs so
that it can be shared

Test Plan: test_jit.py TestJit.test_cpp

gh-metadata: pytorch pytorch 19086 gh/zdevito/13/head

ZolotukhinM commented on aten/src/ATen/core/function_schema.cpp:
```diff
@@ -0,0 +1,142 @@
+void FunctionSchema::findErrorInKwargs(const std::vector<std::string>& kwargs) const {
+  // First check if any of the kwargs are unknown, i.e. don't match the name of
+  // any argument in the schema.
+  for (const auto& kwarg : kwargs) {
```
Why not `std::find` like in the loop below? Also, I wonder if it's worth putting both `arguments` and `kwargs` into an `unordered_set`.

zdevito replied:
it needs to be a lambda variant because it is matching argument.name against kwarg. Not sure why it was count_if rather than find_if, I moved it from another file...

ZolotukhinM replied:
Let's keep it as-is since it's just a move then. We can consider cleaning it up later.

ZolotukhinM commented on aten/src/ATen/core/function_schema.cpp:
```diff
@@ -0,0 +1,142 @@
+  }
+}
+
+void FunctionSchema::checkAndNormalizeInputs(std::vector<IValue>& inputs, const std::unordered_map<std::string, IValue> & kwargs) const {
```
Clang-format please :)

ZolotukhinM reviewed (approved):
Looks good!

zdevito committed:
Update on "Support Kwargs in C++ Function/Method calls"

Support Kwargs in C++ Function/Method calls

Summary: Adds a Kwargs optional argument to the operator() for Method
and Function. Refactors logic from the python support for kwargs so
that it can be shared

Test Plan: test_jit.py TestJit.test_cpp

gh-metadata: pytorch pytorch 19086 gh/zdevito/13/head

zdevito committed:
Update on "Support Kwargs in C++ Function/Method calls"

Support Kwargs in C++ Function/Method calls

Summary: Adds a Kwargs optional argument to the operator() for Method
and Function. Refactors logic from the python support for kwargs so
that it can be shared

Test Plan: test_jit.py TestJit.test_cpp

gh-metadata: pytorch pytorch 19086 gh/zdevito/13/head

zdevito committed:
Update on "Support Kwargs in C++ Function/Method calls"

Support Kwargs in C++ Function/Method calls

Summary: Adds a Kwargs optional argument to the operator() for Method
and Function. Refactors logic from the python support for kwargs so
that it can be shared

Test Plan: test_jit.py TestJit.test_cpp

gh-metadata: pytorch pytorch 19086 gh/zdevito/13/head

zdevito committed:
Update on "Support Kwargs in C++ Function/Method calls"

Support Kwargs in C++ Function/Method calls

Summary: Adds a Kwargs optional argument to the operator() for Method
and Function. Refactors logic from the python support for kwargs so
that it can be shared

Test Plan: test_jit.py TestJit.test_cpp

gh-metadata: pytorch pytorch 19086 gh/zdevito/13/head

zdevito committed:
Update on "Support Kwargs in C++ Function/Method calls"

Support Kwargs in C++ Function/Method calls

Summary: Adds a Kwargs optional argument to the operator() for Method
and Function. Refactors logic from the python support for kwargs so
that it can be shared

Test Plan: test_jit.py TestJit.test_cpp

gh-metadata: pytorch pytorch 19086 gh/zdevito/13/head

zdevito committed:
Update on "Support Kwargs in C++ Function/Method calls"

Support Kwargs in C++ Function/Method calls

Summary: Adds a Kwargs optional argument to the operator() for Method
and Function. Refactors logic from the python support for kwargs so
that it can be shared

Test Plan: test_jit.py TestJit.test_cpp

gh-metadata: pytorch pytorch 19086 gh/zdevito/13/head

zdevito committed:
Update on "Support Kwargs in C++ Function/Method calls"

Support Kwargs in C++ Function/Method calls

Summary: Adds a Kwargs optional argument to the operator() for Method
and Function. Refactors logic from the python support for kwargs so
that it can be shared

Test Plan: test_jit.py TestJit.test_cpp

gh-metadata: pytorch pytorch 19086 gh/zdevito/13/head

zdevito committed:
Update on "Support Kwargs in C++ Function/Method calls"

Support Kwargs in C++ Function/Method calls

Summary: Adds a Kwargs optional argument to the operator() for Method
and Function. Refactors logic from the python support for kwargs so
that it can be shared

Test Plan: test_jit.py TestJit.test_cpp

gh-metadata: pytorch pytorch 19086 gh/zdevito/13/head

zdevito committed:
Update on "Support Kwargs in C++ Function/Method calls"

Support Kwargs in C++ Function/Method calls

Summary: Adds a Kwargs optional argument to the operator() for Method
and Function. Refactors logic from the python support for kwargs so
that it can be shared

Test Plan: test_jit.py TestJit.test_cpp

gh-metadata: pytorch pytorch 19086 gh/zdevito/13/head
